adding a sign up screen and creating animations

shrevin opened the pull request:
created animations in login screen and sign up screen and created password constraints using regex

Commit:
adding a sign up screen and creating animations

created animations in login screen and sign up screen and created password constraints using regex

shrevin commented:
![ezgif com-gif-maker](https://user-images.githubusercontent.com/36344124/181821674-b5dd0794-aacd-4274-ab1f-a265ef0eb418.gif)

arthpatel29 commented on Splash/Splash/Helper/Helper.m:
```diff
@@ -20,4 +21,47 @@ + (NSString*) formatTimeString:(int)secs {
+}
+
++ (void) alertEmptyEmail {
+    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Missing Email" message:@"Please enter an email" preferredStyle:(UIAlertControllerStyleAlert)];
+        // create an OK action
+        UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK"
+                                                           style:UIAlertActionStyleDefault
+                                                         handler:^(UIAlertAction * _Nonnull action) {
+                                                                 // handle response here.
+                                                         }];
+        // add the OK action to the alert controller
+        [alert addAction:okAction];
+        [[UIApplication sharedApplication].keyWindow.rootViewController presentViewController:alert animated:YES completion:^{}];
```
This code is repeated 3 times, we can put all of them into one helper method and use it

Commit:
adding breakpoints file to gitignore

Commit:
Added a Dependency Inversion

This commit does not change any functionality in the app, but adds a dependency inversion where Parse backend depends on the view controllers. There is a singleton for Parse network calls, and this class implements the DataLoaderProtocol containing all the methods needed to access data from Parse

arthpatel29 commented on Splash/Splash/Details/DetailsViewController.m:
```diff
@@ -24,6 +26,7 @@ @interface DetailsViewController () <UITableViewDelegate, UITableViewDataSource,
 @property (strong, nonatomic) PFUser *current;
 @property (strong, nonatomic) IBOutlet UIButton *onLBButton;
 @property (strong, nonatomic) NSArray *friends;
+@property ParseDataLoaderManager *dataLoader;
```
This should be id<DataLoaderProtocol> as discussed earlier.

arthpatel29 commented on Splash/Splash/Details/DetailsViewController.m:
```diff
@@ -50,10 +53,10 @@ - (void) setupTableView {
 }
 
 - (void) setupView {
-    self.usernameLabel.text = self.user[@"username"];
```
Divide this into smaller functions

arthpatel29 commented on Splash/Splash/Networking/ParseDataLoaderManager.m:
```diff
@@ -0,0 +1,183 @@
+
+- (void) loginUser: (NSString *)username password:(NSString *)password vc:(UIViewController *) vc segueId:(NSString *)segueId alert:(UIAlertController *)alert{
+    [PFUser logInWithUsernameInBackground:username password:password block:^(PFUser * user, NSError *  error) {
+        if (error != nil) {
+            DLog(@"User log in failed: %@", error.localizedDescription);
+            [[UIApplication sharedApplication].keyWindow.rootViewController presentViewController:alert animated:YES completion:^{}];
+        } else {
+            DLog(@"User logged in successfully");
+            [vc performSegueWithIdentifier:@"segueFromLogin" sender:self];
+            // display view controller that needs to shown after successful login
+        }
```
Put this inside completion handler

arthpatel29 commented on Splash/Splash/Networking/DataLoaderProtocol.h:
```diff
@@ -0,0 +1,41 @@
+NS_ASSUME_NONNULL_BEGIN
+
+@protocol DataLoaderProtocol 
+- (void) registerUser: (NSString *)username password:(NSString *)password email:(NSString *)email;
+- (void) loginUser: (NSString *)username password:(NSString *)password vc:(UIViewController *) vc segueId:(NSString *)segueId alert:(UIAlertController *)alert;
+- (void) logout: (UIViewController *)vc;
+- (void) getProfileImage: (PFImageView *)imageView user:(PFUser *)user;
+- (NSString *) getGoal: (PFUser *)user;
+- (int) getBubblescore: (PFUser *)user;
+- (int) getStreak: (PFUser *)user;
+- (int) getNumShowers: (PFUser *)user;
+- (int) getTotalShowerTime: (PFUser *)user;
+- (NSString *) getUsername: (PFUser *)user;
+- (void) changeProfileImage: (NSData *)imageData;
+- (void) addFriend: (PFUser *)user;
+- (void) removeFriend: (PFUser *)user;
+- (PFUser *) getCurrentUser;
+- (void) updateGoal: (NSDate *)goal;
+- (NSMutableArray *) getLeaderboardData;
+- (NSDate *) getGoalAsDate: (PFUser *) user;
+- (void) updateBubblescore: (PFUser *) user newScore:(int)newScore;
+- (void) updateStreak: (PFUser *) user newStreak:(int)newStreak;
+- (void) updateTotalShowerTime: (PFUser *) user newTime:(int)newTime;
+- (void) updateNumShowers: (PFUser *) user newNum:(int)newNum;
+- (NSArray *) getFriends: (PFUser *) user;
+- (NSArray *) getPossibleNewFriends: (NSArray*)currFriends;
+- (void) getShowerData: (NSMutableArray *) originalArray filteredArray:(NSMutableArray *)filteredArray tableView:(UITableView *)tableView;
```
Lets group together similar functions

arthpatel29 commented on Splash/Splash/Networking/ParseDataLoaderManager.m:
```diff
@@ -0,0 +1,183 @@
+
+// getting the current user's goal stat
+- (NSString *) getGoal: (PFUser *)user {
+    NSCalendar *calendar = [NSCalendar currentCalendar];
+    NSDateComponents *components = [calendar components:(NSCalendarUnitMinute | NSCalendarUnitSecond) fromDate:[PFUser currentUser][@"goal"]];
+    int currMinute = [components minute];
+    int currSeconds = [components second];
+    return [NSString stringWithFormat:@"%@", [Helper formatTimeString:(currMinute*60) + currSeconds]];
```
Lets refactor this.

Commit:
Finishing Parse Singleton

In this commit, I implemented feedback such as dividing large amounts of code into smaller functions and creating completion handlers for some of the functions in the ParseDataLoader Manager. I also finished the dependency inversion for all of the view controllers.

arthpatel29 commented on Splash/Splash/Login/LoginViewController.m:
```diff
@@ -85,27 +77,38 @@ - (void) changeImage {
     NSString *username = self.usernameField.text;
     NSString *password = self.passwordField.text;
-    [self.dataLoader loginUser:username password:password vc:self segueId:@"segueFromLogin" alert:alert];
+    [self.dataLoader loginUser:username password:password completion:^(PFUser *user, NSError *error) {
```
Is user parameter needed in the completion

arthpatel29 commented on Splash/Splash/Routine/RoutineViewController.swift:
```diff
@@ -0,0 +1,182 @@
+        // Add the subview to the alert's UI property
+        alert.customSubview = subview
+        _ = alert.addButton("Add") {
+            print("Clicked Add")
+            print(textfield1.text)
+            print(textfield2.text)
+            print(textfield3.text)
```
let's remove print

arthpatel29 commented on Splash/Splash/Routine/RoutineViewController.swift:
```diff
@@ -0,0 +1,182 @@
+        let x = (subview.frame.width - 180) / 2
+                
+        // Add textfield 1
+        let textfield1 = UITextField(frame: CGRect(x: x,y: 10,width: 180,height: 35))
+        textfield1.layer.borderColor = UIColor.black.cgColor
+        textfield1.layer.borderWidth = 1.5
+        textfield1.layer.cornerRadius = 5
+        textfield1.placeholder = "Title"
+        textfield1.textAlignment = NSTextAlignment.center
+        subview.addSubview(textfield1)
+                
+        // Add textfield 2
+        let textfield2 = UITextField(frame: CGRect(x: x,y: textfield1.frame.maxY + 10,width: 180,height: 35))
+        textfield2.layer.borderWidth = 1.5
+        textfield2.layer.cornerRadius = 5
+        textfield2.layer.borderColor = UIColor.black.cgColor
+        textfield2.placeholder = "# of minutes"
+        textfield2.textAlignment = NSTextAlignment.center
+        textfield2.textContentType = .dateTime;
+        subview.addSubview(textfield2)
+        
+        // Add textfield 3
+        let textfield3 = UITextField(frame: CGRect(x: x,y: textfield2.frame.maxY + 10,width: 180,height: 35))
+        textfield3.layer.borderWidth = 1.5
+        textfield3.layer.cornerRadius = 5
+        textfield3.layer.borderColor = UIColor.black.cgColor
+        textfield3.placeholder = "# of seconds"
+        textfield3.textAlignment = NSTextAlignment.center
+        subview.addSubview(textfield3)
```
We can move this repeated code to a helper method

yaxuanwang commented on Splash/Splash/Details/DetailsViewController.m:
```diff
@@ -24,6 +26,7 @@ @interface DetailsViewController () <UITableViewDelegate, UITableViewDataSource,
 @property (strong, nonatomic) PFUser *current;
 @property (strong, nonatomic) IBOutlet UIButton *onLBButton;
 @property (strong, nonatomic) NSArray *friends;
+@property id <DataLoaderProtocol> dataLoader;
```
Shall we add attributes for this property?

yaxuanwang commented on Splash/Splash/Friends/FriendsCell.h:
```diff
@@ -8,13 +8,16 @@
 @property (strong, nonatomic) IBOutlet UILabel *friendName;
 @property (strong, nonatomic) IBOutlet PFImageView *friendPic;
 @property (strong, nonatomic) PFUser *user;
+@property ParseDataLoaderManager *dataLoader;
```
same here

yaxuanwang commented on Splash/Splash/Friends/FriendsViewController.m:
```diff
@@ -17,19 +19,28 @@ @interface FriendsViewController () <UICollectionViewDelegate, UICollectionViewD
 - (void)viewDidLoad {
     [super viewDidLoad];
     // Do any additional setup after loading the view.
+    self.dataLoader = [[ParseDataLoaderManager alloc] init];
```
nit: just use [ParseDataLoaderManager new]?

yaxuanwang commented on Splash/Splash/Impact/ImpactViewController.m:
```diff
@@ -54,15 +54,13 @@ - (void) dismissKeyboard
-//    ImpactCalculationsViewController *impactCalc = [segue destinationViewController];
-//    impactCalc.waterFlow = [self.waterFlowTextField.text floatValue];
-//    impactCalc.showersPerWeek = [self.showersTextField.text intValue];
 }
```
Do you still need this function?

yaxuanwang commented on Splash/Splash/Routine/RoutineCell.m:
```diff
@@ -0,0 +1,12 @@
+
+#import "RoutineCell.h"
+
+@implementation RoutineCell
```
Do you plan to add more implementation here as follow up?

shrevin replied:
Yes, I have.

Commit:
Creating Routine View Controller and Reorganizing code

In this commit, I created the routine view controller where users can add steps in their shower, drag and drop them around, and delete steps. I also incorporated the Swift library SCLAlertView for custom alert views when users enter routine information

shrevin commented:
![routine feature](https://user-images.githubusercontent.com/36344124/182963860-2dd5f8f0-06e1-425b-8ae5-36841c771771.gif)
